x/smart/dns.go

```diff
@@ -0,0 +1,217 @@
+// and domain search.
+func makeFullyQualified(domain string) string {
+	if len(domain) > 0 && domain[len(domain)-1] == '.' {
+		return domain
+	}
+	return domain + "."
+}
+
+// mixCase randomizes the case of the domain letters.
+func mixCase(domain string) string {
+	var mixed []rune
+	for _, r := range domain {
+		if rand.Intn(2) == 0 {
+			mixed = append(mixed, unicode.ToLower(r))
+		} else {
+			mixed = append(mixed, unicode.ToUpper(r))
+		}
+	}
+	return string(mixed)
+}
+
+func evaluateNetResolver(ctx context.Context, resolver *net.Resolver, testDomain string) ([]net.IP, error) {
+	requestDomain := mixCase(testDomain)
+	_, err := lookupCNAME(ctx, requestDomain)
+	if err != nil {
+		return nil, fmt.Errorf("could not get cname: %w", err)
+	}
+	ips, err := resolver.LookupIP(ctx, "ip", requestDomain)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup IPs: %w", err)
+	}
+	if len(ips) == 0 {
+		return nil, fmt.Errorf("no ip answer")
+	}
+	for _, ip := range ips {
+		if ip.IsLoopback() {
+			return nil, fmt.Errorf("localhost ip: %v", ip) // -1
+		}
+		if ip.IsPrivate() {
+			return nil, fmt.Errorf("private ip: %v", ip) // -1
+		}
+		if ip.IsUnspecified() {
+			return nil, fmt.Errorf("zero ip: %v", ip) // -1
+		}
+		// TODO: consider validating the IPs: fingerprint, hardcoded ground truth, trusted response, TLS connection.
+	}
+	return ips, nil
+}
+
+func getIPs(answers []dnsmessage.Resource) []net.IP {
+	var ips []net.IP
+	for _, answer := range answers {
+		if answer.Header.Type != dnsmessage.TypeA && answer.Header.Type != dnsmessage.TypeAAAA {
+			continue
+		}
+		var ip net.IP
+		switch rr := answer.Body.(type) {
+		case *dnsmessage.AResource:
+			ip = net.IP(rr.A[:])
+		case *dnsmessage.AAAAResource:
+			ip = net.IP(rr.AAAA[:])
+		default:
+			continue
+		}
+		ips = append(ips, ip)
+	}
+	return ips
+}
+
+func evaluateAddressResponse(response dnsmessage.Message, requestDomain string) ([]net.IP, error) {
+	if response.RCode != dnsmessage.RCodeSuccess {
+		return nil, fmt.Errorf("rcode is not success: %v", response.RCode)
+	}
+	if len(response.Answers) == 0 {
+		return nil, errors.New("no answers")
+	}
+	ips := getIPs(response.Answers)
+	if len(ips) == 0 {
+		return ips, fmt.Errorf("no ip answer: %v", response.Answers)
+	}
+	for _, ip := range ips {
+		if ip.IsLoopback() {
+			return nil, fmt.Errorf("localhost ip: %v", ip)
+		}
+		if ip.IsPrivate() {
+			return nil, fmt.Errorf("private ip: %v", ip)
+		}
+		if ip.IsUnspecified() {
+			return nil, fmt.Errorf("zero ip: %v", ip)
+		}
+	}
+	// All popular recursive resolvers we tested maintain the domain case of the request.
+	// Note that this is not the case of authoritative resolvers. Some of them will return
+	// a fully normalized domain name, or normalize part of it.
+	if response.Answers[0].Header.Name.String() != requestDomain {
+		return ips, fmt.Errorf("domain mismatch: got %v, expected %v", response.Answers[0].Header.Name, requestDomain)
```

Comment:
Just double check, do we return the non-`nil` `ips` or do we need to return `nil, fmt.Errorf(...)`?

Reply:
We return the IPs because we have them. It's helpful for debugging.